### PR TITLE
Update: allow template literals (fixes #5234)

### LINKED ISSUE #5234: Allow enforcing single/double-quotes without preventing use of template literals
How it worked in ESLint 1.0.0 worked perfectly. I could enforce single-quotes everywhere `'quotes': [2, 'single']` and still use template literals when convenient, e.g. if the content had single-quotes so not having to escape it. With ESLint 2.0.0 template literals are now awkwardly forced into the rule and there's no way to enforce single-quotes without also preventing use of template literals when they're not strictly required. I don't want to use double-quotes to prevent having to escape strings containing single-quotes. I want to use template literals, or rather, I already do, and this change makes everything break for me...

I've read both https://github.com/eslint/eslint/issues/5147 and https://github.com/eslint/eslint/issues/3090, but I disagree with the conclusion and I think they're missing this use-case.

To be clear, this used to pass:

```
~/dev
❯ eslint --version                                                                           
v2.0.0
                                                                                                                                               
~/dev
❯ echo "var foo = \`bar 'foo' baz\`;" | eslint --stdin --env=es6 --rule='quotes: [2, single]'

<text>
  1:11  error  Strings must use singlequote  quotes

✖ 1 problem (1 error, 0 warnings)
```

I would like an option to allow this functionality.

### PR DESCRIPTION
**Pull request by BarryThePenguin**
- [x] documentation
- [x] tests

fixes #5234

### CODE REVIEW

**Comment by nzakas** [comment]
@BarryThePenguin this looks like a good start - are you planning on finishing up?

**Comment by BarryThePenguin** [comment]
I've got documentation in the works, it's proving difficult though. I'll commit what I have soon, though it won't be complete. 

Any advice on how to flag the old way as being deprecated? How does eslint handle that?

**Comment by nzakas** [comment]
We just handle it in the documentation. Basically, you write up the new options as if they are the only way of doing it, and then in a small section underneath that, mention the deprecated way of doing things.

**Comment by BarryThePenguin** [comment]
I've added the new options to the documentation

**Comment by ilyavolodin** [comment]
I think this needs a bit more tests. I would like to test some permutations of quote types + allowTemplateLiterals and some negative tests as well.

**Comment by BarryThePenguin** [comment]
Updated with some negative tests. Also changes the tests over to use `{ avoidEscape: true }` as I'd missed that previously...

**Comment by nzakas** [comment]
This looks good. The one thing that jumps out is that we still need to mention `avoid-escape` in the docs. It should say just that this is a deprecated syntax and that you should use the object form instead.

**Comment by BarryThePenguin** [comment]
Updated - as requested :space_invader:

File: docs/rules/quotes.md [hunk 1]
```diff
@@ -23,17 +23,20 @@ This rule is aimed at ensuring consistency of string quotes and as such will rep
...
+When using `"single"` or `"double"`, template literals that don't contain a substitution, don't contain a line break and aren't tagged templates, are flagged as problems, even with the `"avoidEscape"` option. However they are not problems when `"allowTemplateLiterals"` is used.
 
 Configuration looks like this:
 
 ```js
-[2, "single", "avoid-escape"]
+[2, "single", {"avoidEscape": true, "allowTemplateLiterals": true}]
 ```
 
+**Deprecation notice**: The `avoid-escape` enum is a deprecated syntax and you should use the object form instead.
```
Comment (nzakas) [inline docs/rules/quotes.md]: Small change here: change "enum" to "option" (it's not an enumeration)

**Comment by BarryThePenguin** [comment]
Done and rebased

**Comment by nzakas** [comment]
LGTM. Thanks so much for your patience and hard work on this.

**Comment by BarryThePenguin** [comment]
Cheers!

**Pull request merged**